## Test evidence

Run on Node 20.20.2 (`.nvmrc`), pnpm 10.11.0, `packages/core`, vitest 4.1.2.

**RED** — tests against unmodified steps (`git stash push -- packages/core/src/steps/compress.ts packages/core/src/steps/exif.ts`). The assertion diff *is* the flatten repro: an 85-byte `image/gif` comes back as a 9-byte `image/jpeg`.

```
 FAIL  tests/steps/image-processing.test.ts >
       animated images skip the canvas re-encode steps >
       leaves an animated GIF untouched on the web-worker path too

 AssertionError: expected File { size: 9,  type: 'image/jpeg', name: 'loop.gif' }
                      to be File { size: 85, type: 'image/gif',  name: 'loop.gif' }

   - Expected
   + Received
     File {
       "id": "file-1",
       "metadata": {},
       "source": "local",
       "status": "IDLE",
   -   Symbol(kLength): 85,
   -   Symbol(kType): "image/gif",
   +   Symbol(kLength): 9,
   +   Symbol(kType): "image/jpeg",
     }

 Test Files  1 failed | 1 passed (2)
      Tests  7 failed | 27 passed (34)
```

The 7 failures are the animated GIF / animated WebP / APNG cases across both
`compressStep` and `exifStep`, plus the worker-path case. The two that pass in
the RED run are the ones that must *not* change: the still-GIF re-encode and
the animated-GIF thumbnail.

**GREEN** — same two files with the guard restored:

```
 Test Files  2 passed (2)
      Tests  34 passed (34)
```

**Full gates** (all via `rtk proxy` + raw exit codes, per CLAUDE.md's machine notes):

| Gate | Result |
| --- | --- |
| `vitest run` (packages/core) | 143 files, 1722 passed — exit 0 |
| `tsc --noEmit` | exit 0 |
| `tsc -p tsconfig.test.json --noEmit` | exit 0 |
| `eslint . --max-warnings 0` (core) | exit 0 |
| `prettier --check` (all 9 packages' src) | exit 0 |
| `test:quality` | 392 test files + 5 workflows clean, 0 exceptions |
| `size-limit` | all 9 budgets pass — `@upupjs/core` **403.25 kB / 410 kB** |
| `@upupjs/server` test | 32 files, 337 passed, 6 skipped |
| pre-push (`typecheck` + `turbo lint` + `knip`) | exit 0 |

The size line is the one worth a glance: the module is small and dependency-free,
and it rides the existing lazy `import()` of `steps/compress` / `steps/exif` in
`buildAutoPipeline`, so core's mandatory path is unchanged.

`pnpm run e2e` was **not** run — it needs Docker MinIO plus a six-storybook
webServer boot and this box was heavily loaded. CI's `E2E` job is the check to
watch. Worth noting that the e2e harness is Chromium-only, so a live check of an
animated GIF surviving `imageCompression` in Firefox/Safari would be the real
proof if you want one before a release.

### One thing I did not fix

`outputTypeFor` sends anything that is not PNG or WebP to `image/jpeg` while
keeping the original filename, so a **still** GIF becomes JPEG bytes still named
`.gif`. That is pre-existing and orthogonal to this PR, so I left it alone —
flagging it in case you want it tracked separately.

--- a/apps/landing/content/docs/guides/file-processing.mdx
+++ b/apps/landing/content/docs/guides/file-processing.mdx
@@ -29,8 +29,8 @@ The pipeline is assembled from boolean options on the uploader. The order is
 | Order | Step        | Option                 | Applies to       |
 | ----- | ----------- | ---------------------- | ---------------- |
 | 1     | `heic`      | `heicConversion`       | HEIC/HEIF images |
-| 2     | `exif`      | `stripExifData`        | any `image/*`    |
-| 3     | `compress`  | `imageCompression`     | any `image/*`    |
+| 2     | `exif`      | `stripExifData`        | still `image/*`  |
+| 3     | `compress`  | `imageCompression`     | still `image/*`  |
 | 4     | `thumbnail` | `thumbnailGenerator`   | any `image/*`    |
 | 5     | `hash`      | `checksumVerification` | every file       |
 
@@ -120,6 +120,11 @@ Two things to know about the mechanism:
   empty type falls back to `image/jpeg`). It does not resize.
 - **It is a re-encode, not a surgical metadata edit.** Everything outside the
   pixel data goes, including ICC color profiles.
+- **Animated images are skipped.** A canvas cannot write an animation, so
+  re-encoding an animated GIF, animated WebP or APNG would keep frame one and
+  discard the rest. Those files are left untouched — which also means their EXIF
+  is left untouched. `imageCompression` skips them for the same reason. If you
+  need metadata stripped from animated uploads, do it server-side.
 
 Because EXIF runs _before_ compression, turning both on means two encodes. If
 you want compression anyway, that is the cost of the privacy guarantee; if you

--- a/apps/landing/content/docs/guides/processing/compression.mdx
+++ b/apps/landing/content/docs/guides/processing/compression.mdx
@@ -76,14 +76,22 @@ does.
 ## Output format
 
 The encoder preserves `image/png` and `image/webp`; **every other input type
-comes out as `image/jpeg`**. The filename is unchanged, so a `.gif` compressed
-to JPEG bytes keeps its original name — set the storage key or rename the file
-yourself if the extension matters to you.
+comes out as `image/jpeg`**. The filename is unchanged, so a still `.gif`
+compressed to JPEG bytes keeps its original name — set the storage key or
+rename the file yourself if the extension matters to you. (An _animated_ GIF is
+not compressed at all — see below.)
 
 ## When compression does nothing
 
-Compression can decide to keep the original file. Two cases:
+Compression can decide to keep the original file. Three cases:
 
+- **The image is animated.** A canvas has no animated encoder, so re-encoding an
+  animated GIF, animated WebP or APNG would upload its first frame and throw the
+  rest away. upup detects those from the file's bytes and skips the step
+  entirely — the original uploads untouched, with no `compressed` metadata. Still
+  images of the same formats are compressed normally. `stripExifData` skips them
+  for the same reason, so an animated image reaches your storage exactly as the
+  user picked it, EXIF included.
 - **No size benefit.** If the re-encoded result is not smaller than the
   original, upup keeps the original — but only when you passed neither
   `maxSizeMB` nor an explicit `maxWidthOrHeight`. Setting either of those is read

